OpenSim/Common/Test/testDataInterface.cpp

```diff
@@ -0,0 +1,375 @@
+        // throw an exception to that effect.
+        ASSERT_THROW(OpenSim::Exception, DataAdapter::createAdapter("stream"));
+
+        auto* adapter = DataAdapter::createAdapter("streamWriter");
+        adapter->prepareForWriting(table);
+        adapter->write();
+
+        // Write out contents to cout
+        cout << endl;
+
+        std::stringstream iobuffer;
+        StreamWriter<SimTK::Vec3>  markerWriter(iobuffer);
+        markerWriter.prepareForWriting(markers);
+        markerWriter.write();
+        cout << iobuffer.str() << endl;
+
+        StreamReader<> reader(iobuffer);
+        DataTable readTable;
+
+        reader.prepareForReading(readTable);
+        reader.read();
+        readTable.getNumRows();
+        size_t nrows = readTable.getNumCols();
+
+        iobuffer.clear();
+        DataTable_<SimTK::Mat22> matTable;
+        matTable.appendRow(RowVector_<Mat22>(3, Mat22(1.0) ) );
+        matTable.appendRow(RowVector_<Mat22>(3, Mat22(2.0) ) );
+        matTable.appendRow(RowVector_<Mat22>(3, Mat22(3.0) ) );
+        matTable.appendRow(RowVector_<Mat22>(3, Mat22(4.0) ) );
+        Array<string> matlabels;
+        matlabels.append("mat_1"); 
+        matlabels.append("mat_2");
+        matlabels.append("mat_3");
+        matTable.updColumnLabels() = matlabels;
+
+        StreamWriter<Mat22>  matWriter(iobuffer);
+        matWriter.prepareForWriting(matTable);
+        matWriter.write();
+        cout << iobuffer.str() << endl;
+
+        TimeSeriesData tsd(table);
+        tsd.dumpToStream(cout);
+
+        // Time column in data table should not be part of the data in
+        // time series data
+        ASSERT((table.getNumCols() - 1) == tsd.getNumCols());
+
+        cout << "\nCopy construct TimeSeriesData" << endl;
+        TimeSeriesData tsd2(tsd);
+        tsd2.dumpToStream(cout);
+
+        DataTable* dt_tsd = &tsd;
+        cout << "\nCopy construct TimeSeriesData from DataTable reference to TimeSeriesData" << endl;
+        TimeSeriesData tsd3(*dt_tsd);
+        tsd3.dumpToStream(cout);
+
+        CSVFileReader csvReader("dataTestFile.csv");
+        ASSERT(csvReader.isWriteAccess() == false);
+        ASSERT(csvReader.isReadAccess() == true);
+        DataTable csvTable;
+        csvReader.prepareForReading(csvTable);
+        csvReader.read();
+
+        cout << endl;
+        cout << "CSV Comments: " << csvTable.getMetaData() << endl;
+        cout << "CSV Labels: " << csvTable.getColumnLabels() << endl;
+        cout << "CSV nrows = " << csvTable.getNumRows();
+        cout << "    ncols = " << csvTable.getNumCols() << endl;
+
+        DataTable csvTable2("dataTestFile.csv");
+        cout << "DataTable csvTable2('dataTestFile.csv')" << endl;
+        csvTable2.dumpToStream(cout);
+
+        // Try to read the table in again. Should throw an exception
+        // about no more lines to read from file.
+        ASSERT_THROW(std::exception, csvReader.read());
```

Comment:
Is there a way to find out if there are more lines to read? Otherwise, the only way for a client to know is by checking for an exception.